src/plugins/discord/mirror.js

```diff
@@ -0,0 +1,384 @@
+// @flow
+
+import type Database from "better-sqlite3";
+import stringify from "json-stable-stringify";
+import * as Model from "./models";
+import dedent from "../../util/dedent";
+
+const VERSION = "DISCORD_MIRROR_v1";
+
+/**
+ * Persists a local copy of data from a Discord Guild.
+ * Implements create and read functionality.
+ *
+ * Each Mirror instance is tied to a particular Guild. Trying to use a mirror
+ * for multiple Discord Guilds is not permitted; use separate Mirrors.
+ */
+
+export class Mirror {
+  +_db: Database;
+
+  /**
+   * Construct a new Mirror instance.
+   *
+   * Takes a Database and GuildId.
+   *
+   */
+
+  constructor(db: Database, guildId: Model.Snowflake) {
+    if (db == null) throw new Error("db: " + String(db));
+    this._db = db;
+    this._initialize(guildId);
+  }
+
+  _transaction(queries: () => void) {
+    const db = this._db;
+    if (db.inTransaction) {
+      throw new Error("already in transaction");
+    }
+    try {
+      db.prepare("BEGIN").run();
+      queries();
+      if (db.inTransaction) {
+        db.prepare("COMMIT").run();
+      }
+    } finally {
+      if (db.inTransaction) {
+        db.prepare("ROLLBACK").run();
+      }
+    }
+  }
+
+  _initialize(guild: Model.Snowflake) {
+    const db = this._db;
+    // We store the config in a singleton table `meta`, whose unique row
+    // has primary key `0`. Only the first ever insert will succeed; we
+    // are locked into the first config.
+    db.prepare(
+      dedent`\
+        CREATE TABLE IF NOT EXISTS meta (
+            zero INTEGER PRIMARY KEY,
+            config TEXT NOT NULL
+        )
+      `
+    ).run();
+
+    const config = stringify({
+      version: VERSION,
+      guild,
+    });
+
+    const existingConfig: string | void = db
+      .prepare("SELECT config FROM meta")
+      .pluck()
+      .get();
+    if (existingConfig === config) {
+      // Already set up; nothing to do.
+      return;
+    } else if (existingConfig !== undefined) {
+      throw new Error(
+        "Database already populated with incompatible server or version"
+      );
+    }
+    db.prepare("INSERT INTO meta (zero, config) VALUES (0, ?)").run(config);
+
+    const tables = [
+      dedent`\
+        CREATE TABLE channels (
+            id TEXT PRIMARY KEY,
+            type TEXT NOT NULL,
+            name TEXT NOT NULL
+        )
+      `,
+      dedent`\
+        CREATE TABLE members (
+            user_id TEXT PRIMARY KEY,
+            username TEXT NOT NULL,
+            discriminator TEXT NOT NULL,
+            bot INTEGER NOT NULL,
+            nick TEXT
+        )
```

Comment:
This seems like a nice approach (yay correctness). If we do this, let's of course add test cases too.

Minor nit on your `toMember` function: it would be much cleaner, imo, to return `Member | null` rather than throwing an error. Let's the caller decide what to do in such cases, rather than forcing the caller to check first. (If the caller has to check first to avoid a runtime error, then there isn't any real benefit to the function, since once the caller has checked, they already know they have a member.)

Reply:
@Beanow I'm not sure what value it adds to bend the model to accommodate Users that are no longer members. What is the downside to treating them like Members, or how are we using that info to our advantage? Bear in mind, we could also have GuildMembers in our tables that subsequently leave and become non-members. Why not just add a boolean field on GuildMember that distinguishes whether a User is a `CurrentMember`? Also bear in mind, as long as Discord's API exposes a Members route, it will be easy to check whether the members in our table are current members. 

*I'm working under the assumption that we'll pick up these Non-Member Users through Mentions or Reactions, and get a UserId from Discord for those Users.

Reply:
> what value it adds to bend the model to accommodate Users that are no longer members

Currently the hackathon Discord plugin skips users that aren't members, but do show up, maybe because they've sent messages, added reactions or are mentioned. This means, users who _should be_ in the graph, are not. By supporting non-member users we can include them in the graph, as they should be :]

> Why not just add a boolean field on GuildMember that distinguishes whether a User is a CurrentMember?

The membership has some guild-specific data. The nickname and roles included there only apply to that specific guild. While user data like the discriminator and bot boolean are global. So having nullable membership data and user data seems like the model that more accurately reflects this.

> We could have GuildMembers in our tables that subsequently leave and become non-members

Yes, by deleting their membership in the DB but not their user data, it would be up-to-date.

> As long as Discord's API exposes a Members route, it will be easy to check whether the members in our table are current members,

Indeed, for that we can add/remove/update member records (and should, for details like nicknames getting updated).

> so it's not like we can't adjust the shape of the models in the future if a use case emerges, without re-scraping the Discord server or anything; we can choose a simpler model now without bad repercussions if we need to change it later.

This use-case already emerged in the prototype. Non-members appeared in the data and we ignored them with a warning, because there were only a few instances, that was a hack to move forward.

> *I'm working under the assumption that we'll pick up these Non-Member Users through Mentions or Reactions, and get a UserId from Discord for those Users.

Yes, I think that will reveal the user IDs to us. However we need to resolve those to a User object and store it in the Mirror if we want to add them to the Graph. For both display purposes (so we can show `@Beanow#5887` rather than `@<!143776454050709505>`) and things like identifying bots so they can have different weights.

Reply:
@Beanow How did you discover that these Users were not Members? 

I can get behind separate tables for Users and Members and am working on implementing it; however,  it's difficult for me to see how the first model below provides any advantages over the second. 

```js
export type User = {|
  +id: Snowflake,
  +username: string,
  +discriminator: string,
  +bot: boolean,
|};

export type GuildMember = {|
  +user: User,
  +nick: string | null,
|}
```

```js
export type User = {|
  +id: Snowflake,
  +username: string,
  +discriminator: string,
  +bot: boolean,
  +nick: string | null,
  +member: boolean, 
|}
```

Reply:
@BrianLitwin by first mirroring the members, any user ID that is relevant (wrote messages, added reactions, was mentioned, etc.) but not in the user table yet is probably a non-member user we should get more data on separately.

I'm guessing there isn't a convenient single query for this in the API, so we'll probably need to get this data as we find out what's missing.

To understand the difference in the two types. Imagine what would happen if we wanted to have data for multiple guilds in a single database? The user model including a nick and member boolean would be incorrect as nick and member depend on the guild while the user does not. So we'd be taking on some technical debt there.

Reply:
You're totally right about this @Beanow . We're hewing closer to the Discord API by doing it this way versus just keeping a `membership` boolean or something. Updated.